### DISCUSSION TITLE
saveSourceModel(final SaveModelAction action) - getFileUri is empty

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#647] [Q&A] [answered]
I try to implement a custom SourceModelStorage.

In the loadSourceModel I can get the source file uri with

```java 
    @Override
    public void loadSourceModel(final RequestModelAction action) {
        Map<String, String> options = action.getOptions();
        String uri = options.get("uri");
        ........
    }
```

This is a little bit clumsy but I am fine with this.

When I try to implement the method saveSourceModel I expect that SaveModelAction param provides me with the target file uri

```java
    @Override
    public void saveSourceModel(final SaveModelAction action) {
        Optional<String> uri = action.getFileUri();
        ....
     }
```

Beside the fact that the method parameter has a different layout as in the load method the `getFileUri()` call always returns an empty String Optional. I wonder if this is a bug?

My only solution for now is to store the file path during the `loadSourceModel` method into my custom modelState so that I can reuse it the saveSourceModel

**Comment by planger** [comment, marked as answer]
To make querying the options less clumsy, there is a [`ClientOptionsUtil`](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/utils/ClientOptionsUtil.java) util.

The `SaveModelAction` only contains a `fileUri` if it is a save-as action, that is the model should be saved into a different file than the one that is currently opened. See also https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md#251-savemodelaction

The `ModelState` stores all options of the original `RequestModelAction` in the [client options](https://github.com/eclipse-glsp/glsp-server/blob/39ee6c9aa37f71d989ca0efcfc49ae2f25690b3e/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java#L51). So if `fileUri` is not set in the `SaveModelAction`, you can retrieve the original file URI from the model state.

It might help to look at the rather new [`org.eclipse.glsp.server.emf`](https://github.com/eclipse-glsp/glsp-server/tree/master/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf) package. It contains a module set-up and several default implementations specifically for EMF models, such as the [`EMFSourceModelStorage`](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFSourceModelStorage.java).

**Comment by planger** [comment]
If my answer is fine with you and you agree that no code change is required, I suggest that we convert this into a discussion rather than an isse. Any objections @rsoika?

**Comment by rsoika** [comment]
that is ok for me to to close this issue or move it into discussion.
